### PR TITLE
User Tests

### PR DESCRIPTION
**Pull request by natalie-abreu**
Adds tests for User class and UserServlet. Also adds a few methods to the User class useful for testing and refactors IdHelper

### CODE REVIEW

File: src/main/java/com/google/sps/data/IdToken.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+// Copyright 2019 Google LLC
```
Comment (ekengni) [inline src/main/java/com/google/sps/data/IdToken.java]: delete file

File: src/main/java/com/google/sps/data/MockIdHelper.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.HashMap;
+
+
+public final class MockIdHelper implements IdHelper {
+
+    static final String ID_TOKEN_PARAM = "id_token";
+    static final String NAME = "name";
+    static final String USER_ID = "userId";
+
+    private static Map<String, String> mockPayload = new HashMap<String, String>() {
```
Comment (ekengni) [inline src/main/java/com/google/sps/data/MockIdHelper.java:37]: From the documentation I have read, there are unexpected behaviors when initializing a map this way. I suggest using a function to initialize the HashMap.

File: src/main/java/com/google/sps/data/Verifier.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
+// Copyright 2019 Google LLC
+//
```
Comment (ekengni) [inline src/main/java/com/google/sps/data/Verifier.java]: Delete

File: src/test/java/com/google/sps/UserTest.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  @Test
+  public void userNotEqualsTest(){
+    User other = new User.Builder(USER_ID)
+        .setNickname(NICKNAME)
+        .setCurrentChallengeId(DEFAULT_CHALLENGE_ID)
+        .setCreatedEvents(new ArrayList<Long>(Arrays.asList(1L)))
+        .setChallengeStatuses(DEFAULT_CHALLENGE_STATUSES)
+        .build();
+
+        Assert.assertTrue(!user.equals(other));
```
Comment (ekengni) [inline src/test/java/com/google/sps/UserTest.java:95]: use assertFalse

File: src/main/java/com/google/sps/data/MockIdHelper.java (hunk 2 above)
Comment (lazau-g) [reply src/main/java/com/google/sps/data/MockIdHelper.java:37]: I'm actually curious about that Erick - mind linking me the docs that say this? It's very surprising to me and I'd like to learn more about it too.

File: src/test/java/com/google/sps/UserServletTest.java [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+      when(response.getWriter()).thenReturn(pw);
+      servlet.doGet(request, response);
+      pw.flush();
+
+      JSONObject responseJson = new JSONObject(writer.toString().trim());
+      Assert.assertEquals(userJson.getString(User.ID),
+                          responseJson.getString(User.ID));
+      Assert.assertEquals(userJson.getString(User.NICKNAME),
+                          responseJson.getString(User.NICKNAME));
+    } catch (IOException e) {}
```
Comment (lazau-g) [inline src/test/java/com/google/sps/UserServletTest.java:106]: Is there a reason for adding these empty `catch IOException` blocks? It's prevalent throughout these tests and it seems more straightforward to me to just leave them out and have these exceptions fail the tests.

File: src/test/java/com/google/sps/UserServletTest.java [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+
+    StringWriter writer = new StringWriter();
+    PrintWriter pw = new PrintWriter(writer);
+
+    try {
+      when(response.getWriter()).thenReturn(pw);
+      servlet.doGet(request, response);
+      pw.flush();
+
+      Assert.assertEquals(writer.toString().trim(), "");
```
Comment (lazau-g) [inline src/test/java/com/google/sps/UserServletTest.java:124]: If possible, you should also assert response.getStatus() == SC_NOT_FOUND

File: src/test/java/com/google/sps/UserServletTest.java [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+
+    StringWriter writer = new StringWriter();
+    PrintWriter pw = new PrintWriter(writer);
+
+    try {
+      when(response.getWriter()).thenReturn(pw);
+      servlet.doGet(request, response);
+      pw.flush();
+      
+      Assert.assertEquals(writer.toString().trim(), "");
```
Comment (lazau-g) [inline src/test/java/com/google/sps/UserServletTest.java:144]: Again, assert HTTP status code.

File: src/test/java/com/google/sps/UserServletTest.java (hunk 6 above)
Comment (natalie-abreu) [reply src/test/java/com/google/sps/UserServletTest.java:124]: Couldn't figure out how to get the status from the response, both success and error returned 0. I'll try to look into it again and see if I was missing something.

File: src/test/java/com/google/sps/UserServletTest.java (hunk 5 above)
Comment (natalie-abreu) [reply src/test/java/com/google/sps/UserServletTest.java:106]: Compiler fails with an unreported exception error if they are removed

File: src/main/java/com/google/sps/data/MockIdHelper.java (hunk 2 above)
Comment (ekengni) [reply src/main/java/com/google/sps/data/MockIdHelper.java:37]: http://openjdk.java.net/jeps/269
https://stackoverflow.com/questions/6802483/how-to-directly-initialize-a-hashmap-in-a-literal-way

**Pull request merged**